### PR TITLE
Select last section if parent section doesn't exist

### LINKED ISSUE #60972: Inserter: Pattern Category Selection Removes Pattern Tab
### Description

I believe this is due to the pattern category entering zoom-out mode which selects a root block.

### Step-by-step reproduction instructions

1. Open main Block Inserter
2. Click Pattern Tab
3. Click a Pattern Category
4. Click Block Tab
5. Click Pattern Tab
6. Click Pattern Category
7. Bug: Pattern Category disappears and focus returns to Block tab

### Screenshots, screen recording, code snippet

https://github.com/WordPress/gutenberg/assets/967608/20a4dc7f-c9f7-4bee-953e-9255c5776407

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by draganescu**
## What?

Fixes #60972

## Why?

Selecting a category from the inserter patterns tab puts the editor in zoom out mode. Since in zoom out mode block selection is disabled and we default to "section" selection, we need to manage what is selected.

There is some logic to decide what to select in `__unstableSetEditorMode` which is an action in the block editor store, so when zoom out is activated:

- if there is a section root
  - **if there is a block selected, select the parent section**
  - if there is no block selection, select the last section
- if there is no section root
  - if there is a selected block, select the root parent
- finally,
  - select the "last root block" (*)

The bug in #60972 happens because we didn't consider that a selected block may have an undefined parent section (the root block doesn't).

## How?

Fixes `__unstableSetEditorMode` to also select the last section, if the currently selected block does not have a valid parent section.

## Testing Instructions

0. Create a template with several patterns grouped in a group block whose tag is set to `main`
1. Open main Block Inserter
2. Click Pattern Tab
3. Click a Pattern Category
4. Click Block Tab
5. Click Pattern Tab
6. Click Pattern Category
7. **It should work as expected**

### Testing Instructions for Keyboard

n/a

## Screenshots or screencast 

n/a

(*) "last root block", we expect that in a template we'd have a list of patterns or other grouped blocks, so if they're not in a `main` group they're considered sections themselves.

### CODE REVIEW

**Review by scruffian — APPROVED** [review]
LGTM. Pushed a small refactor.

**Review by Mamaduka — APPROVED** [review]
Thank you, @draganescu! The fix works as expected.

Note: The logic inside the `mode === 'zoom-out'` block is becoming a little bit hard to follow. It might be a time for minor refactoring + adding comments.

**Review by Mamaduka — DISMISSED** [review]
I think I found another issue that seems to be related.

## Steps to reproduce.
1. Open the Site Editor.
2. Open the Inserter > Patterns > Any category.
3. Click on the canvas in Zoomed-out view.
4. Notice that the Insert isn't closed, and the Patterns tab is missing.
5. Toggle inserter.
6. Notice that Media tab is also missing

## Screenshot

https://github.com/WordPress/gutenberg/assets/240569/28353f73-0654-415b-b75a-bb12aaadd662

**Comment by scruffian** [comment]
@Mamaduka good catch, but I think this is a slightly different scenario, so I think we should deal with it in a different PR. I created a new issue to keep track of it: https://github.com/WordPress/gutenberg/issues/61038

**Comment by Mamaduka** [comment]
@scruffian, that sounds good. Could you add details on how it's different from the issue @draganescu resolved?
